File: 2017/12-Dec/12-7-Native-Getter-Setter/getset.swift
// Author: Andrew Jarombek
// Date: 12/11/2017
// Get and Set in Swift are for native getters and setters.
// They are also referred to as computed properties

class Person {
    var _first: String
    var _last: String
    var first: String {
        get {
            print("Accessing First Name")
            return self._first
        }
        set {
            print("Setting First Name")
            self._first = newValue
        }
    }
    var last: String {
        get {
            print("Accessing Last Name")
            return self._last
        }
        set {
            print("Setting Last Name")
            self._last = newValue
        }
    }
    var full: String {
        print("Accessing Full Name")
        return "\(self._first) \(self._last)"
    }
    
    init() {
        _first = ""
        _last = ""
    }
}

var andy = Person()
andy.first = "Andy" // Setting First Name
andy.last = "Jarombek" // Setting Last Name

andy.first // Accessing First Name
andy.last // Accessing Last Name
andy.full // Accessing Full Name

File: 2017/12-Dec/12-7-Native-Getter-Setter/getset.js
// @author Andrew Jarombek
// @since 12/7/2017
// Demonstrate Native Getters and Setters in JavaScript

var person = {
    _first: 'Andy',
    _last: 'Jarombek',
    get first() {
        console.info('Accessing First Name');
        return this._first;
    },
    set first(str) {
        console.info('Setting First Name');
        if (typeof str === 'string' || str instanceof String) {
            var s = str.toLowerCase();
            this._first = s.charAt(0).toUpperCase() + s.slice(1);
        } else {
            throw new TypeError('First Name Must be a String');
        }
    },
    get last() {
        console.info('Accessing Last Name');
        return this._last;
    },
    set last(str) {
        console.info('Setting Last Name');
        if (typeof str === 'string' || str instanceof String) {
            var s = str.toLowerCase();
            this._last = s.charAt(0).toUpperCase() + s.slice(1);
        } else {
            throw new TypeError('Last Name Must be a String');
        }
    },
    get full() {
        console.info('Accessing Full Name');
        // Use ES6 template literals
        return `${this._first} ${this._last}`;
    }
};

// Native Getters allow us to call the getter just as we would any property
console.info(person.first);
console.info(person.last);
console.info(person.full); // Andy Jarombek

try {
	// Native Setters allow us to call the setter just as we would 
    // any property assignment
	person.first = 'JOSEPH';
	person.last = 25 // TypeError
} catch(e) {
	console.error(e);
}

console.info(person.full); // Joseph Jarombek